hop/publish.py

```diff
@@ -19,37 +20,54 @@
 def _add_parser_args(parser):
     cli.add_url_opts(parser)
     parser.add_argument(
-        "gcn", metavar="GCN", nargs="+", help="One or more GCNs to publish.",
+        "message", metavar="MESSAGE", nargs="+", help="One or more messages to publish.",
     )
     cli.add_config_opts(parser)
 
+    parser.add_argument(
+        "-f",
+        "--format",
+        type=str,
+        default="blob",
+        help="Specifies the format of the message, such as 'circular' or 'voevent'. "
+        "Specify 'blob' if sending an unstructured message. Default: 'blob'.",
+    )
 
 def _main(args=None):
-    """Parse and publish GCNs.
+    """Parse and publish messages.
 
     """
+
     if not args:
         parser = argparse.ArgumentParser()
         _add_parser_args(parser)
         args = parser.parse_args()
 
+    # if not args.message:
+    #     raise NameError("Error: no message specified. Specify one or more messages to publish.")
+        
     # load config if specified
     config = cli.load_config(args)
 
+    # get format of the message(s) if specified
+    message_format = "blob" if not args.format else args.format
+
     stream = Stream(format="json", config=config)
     with stream.open(args.url, "w") as s:
-        for gcn in args.gcn:
-            extension = os.path.splitext(gcn)[1]
-
-            # parse based on GCN type
-            if extension == ".gcn3":
-                with open(gcn, "r") as f:
-                    gcn = GCNCircular.from_email(f)
-            elif extension == ".xml":
-                with open(gcn, "rb") as f:
-                    gcn = VOEvent.from_xml(f)
+        for message_file in args.message:
+
+            if message_format == "circular":
+                with open(message_file, "r") as f:
+                    message_model = GCNCircular.from_email(f)
+            elif message_format == "voevent":
+                with open(message_file, "rb") as f:
+                    message_model = VOEvent.from_xml(f)
+            elif message_format == "blob":
+                with open(message_file, "r") as f:
+                    message_model = message_blob.from_file(f)
             else:
-                raise ValueError(f"File extension {extension} not recognized")
+                warnings.warn("Warning: format not recognized. Sending as unstructured blob")
+                with open(message_file, "r") as f:
+                    message_model = message_blob.from_file(f)
```

Comment:
One way to get around the `"rb"` vs `"r"` file mode differences is for each function in the `model_loader` map to take a filename, so like 
```python
formats = {
    "circular": GCNCircular.from_email_file,
    "voevent": VOEvent.from_xml_file,
    "blob": message_blob.from_file,
}
...
if message_format in formats:
    loader = formats[message_format]
else:
    # raise the warning, etc
for message_file in args.message:
    message_model = loader(message_file)
```
```python
class VOEvent:
   ...
    @classmethod
    def from_xml_file(cls, filename):
        with open(filename, "rb") as f:
            return cls.from_xml(f)
```

Reply:
Thanks for the pointers. I implemented the mapping with @spenczar's structure so that the file loading can now happen from `models.py`. (I also used the mapping structure in a similar setup to generalize the unit tests)